images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go

```diff
@@ -0,0 +1,340 @@
+	if current.Status.Stats != nil {
+		stats = *current.Status.Stats
+	}
+	pt := NewPhaseTransitions(stats.PhasesTransitions, current.Status.Phase, changed.Status.Phase)
+	stats.PhasesTransitions = pt.Slice
+
+	var launchTimeDuration virtv2.VirtualMachineLaunchTimeDuration
+
+	pending := pt.Map[virtv2.MachinePending]
+	starting := pt.Map[virtv2.MachineStarting]
+	running := pt.Map[virtv2.MachineRunning]
+
+	if !starting.Timestamp.IsZero() && !pending.Timestamp.IsZero() && starting.Timestamp.Time.After(pending.Timestamp.Time) {
+		launchTimeDuration.WaitingForDependencies = metav1.Duration{Duration: starting.Timestamp.Sub(pending.Timestamp.Time)}
+	}
+	if !running.Timestamp.IsZero() && !starting.Timestamp.IsZero() && running.Timestamp.Time.After(starting.Timestamp.Time) {
+		launchTimeDuration.VirtualMachineStarting = metav1.Duration{Duration: running.Timestamp.Sub(starting.Timestamp.Time)}
+	}
+	var empty virtv1.VirtualMachineInstanceGuestOSInfo
+	if kvvmi != nil && empty == current.Status.GuestOSInfo && empty != kvvmi.Status.GuestOSInfo && !running.Timestamp.IsZero() {
+		launchTimeDuration.GuestOSAgentStarting = metav1.Duration{Duration: time.Since(running.Timestamp.Time)}
+	}
+
+	stats.LaunchTimeDuration = launchTimeDuration
+	changed.Status.Stats = &stats
+}
+
+func (h *LifeCycleHandler) wrapMigrationState(state *virtv1.VirtualMachineInstanceMigrationState) *virtv2.VirtualMachineMigrationState {
+	if state == nil {
+		return nil
+	}
+	return &virtv2.VirtualMachineMigrationState{
+		StartTimestamp: state.StartTimestamp,
+		EndTimestamp:   state.EndTimestamp,
+		Target: virtv2.VirtualMachineLocation{
+			Node: state.TargetNode,
+			Pod:  state.TargetPod,
+		},
+		Source: virtv2.VirtualMachineLocation{
+			Node: state.SourceNode,
+		},
+		Result: h.getResult(state),
+	}
+}
+
+func (h *LifeCycleHandler) getResult(state *virtv1.VirtualMachineInstanceMigrationState) virtv2.MigrationResult {
+	if state == nil {
+		return ""
+	}
+	switch {
+	case state.Completed && !state.Failed:
+		return virtv2.MigrationResultSucceeded
+	case state.Failed:
+		return virtv2.MigrationResultFailed
+	default:
+		return ""
+	}
+}
+
+var phasePreviousPhase = map[virtv2.MachinePhase]virtv2.MachinePhase{
+	virtv2.MachineRunning:  virtv2.MachineStarting,
+	virtv2.MachineStarting: virtv2.MachinePending,
+	virtv2.MachineStopped:  virtv2.MachineStopping,
+}
+
+var phases = []virtv2.MachinePhase{
+	virtv2.MachinePending,
+	virtv2.MachineStarting,
+	virtv2.MachineRunning,
+	virtv2.MachineMigrating,
+	virtv2.MachinePause,
+	virtv2.MachineDegraded,
+	virtv2.MachineStopping,
+	virtv2.MachineStopped,
+	virtv2.MachineTerminating,
+}
+
+type PhaseTransitions struct {
+	Slice []virtv2.VirtualMachinePhaseTransitionTimestamp
+	Map   map[virtv2.MachinePhase]virtv2.VirtualMachinePhaseTransitionTimestamp
+}
+
+func NewPhaseTransitions(phaseTransitions []virtv2.VirtualMachinePhaseTransitionTimestamp, oldPhase, newPhase virtv2.MachinePhase) PhaseTransitions {
```

Comment:
It's hard to understand what's going on here and why it's needed. Let's add a comment?